Omniauth developer strategy

KitaitiMakoto opened the pull request:
@kentaro san said "I want you to do with omniauth `developer` mode." to me at issue #29.
I'v done it at RubyWorld Conference 2012.
## Usage:
1. Create a user anyway.
2. Click "Dev sign in" at `/caveat`.
3. Enter the user's name and uid, then push "sign in" button.
## Required Considerations
- POST request to `/auth/developer/callback` is accepted. Is it OK?
- No test. Is it needed even though `developer` strategy is activated only when `RAILS_ENV=development`?

Thanks!

KitaitiMakoto committed:
[REFACTORING]Use dot accessor instead of hash accessor for auth parameters

KitaitiMakoto committed:
WIP: add developer strategy of OmniAuth

KitaitiMakoto committed:
Change fields for developer strategy: name -> nickname, add uid

KitaitiMakoto committed:
Add links to sign in at development environment

KitaitiMakoto committed:
[BUG FIX]:github -> "github"

KitaitiMakoto committed:
Change spec after implementation

kentaro commented on spec/support/signin.rb:
```diff
@@ -1,5 +1,5 @@
 def auth_params_for (user)
-  {
+  Hashie::Mash.new({
```
`Gemfile`に依存かかかなくてもだいじょうぶですか?

kentaro replied:
omniauthの依存で入ってくるんですねー

KitaitiMakoto replied:
そうなんです。
実装の方でも使っていて、OmuniAuthが使うのをやめた時はこけてほしいので書かないままがいいと思います。
便利なので他で使うなら勿論入れちゃっていいと思いますが。

kentaro replied:
了解ですー

banyan replied:
何かのライブラリで依存しているものを暗黙的に利用すると見通しが悪くなる気がするのですが、そうでもないのですかね？

kentaro replied:
そうかもしれぬ

hsbt replied:
ここ、auth_params_for(:foo).uid みたいにすることが無いなら Hash でも良いような。

kentaro replied:
`Hasie::Mash`を使いたいという強い意図を感じたのでそのままにしてありますｗ

KitaitiMakoto replied:
あ、俺か。遅れてすんません。

`Hashie::Mash`を使いたいという強い意図のみなもとはここです。
https://github.com/kentaro/triglav/pull/41/files#L0R14

developer strategyでは`auth_params`に`update_image(auth_params['extra']['raw_info']['avatar_url'])`という物が入ってこないので、

``` ruby
update_image(auth_params['extra'] && auth_params['extra']['raw_info'] && auth_params['extra']['raw_info']['avatar_url'])
```

といったことをしなくちゃいけなくてだるいなあと思ってドットでつなげて`try`とか`present?`とかを使えるようにしたかったです。
（その他の場所でドットを使うようにしたのはほとんどおまけです）

で、

> 何かのライブラリで依存しているものを暗黙的に利用すると見通しが悪くなる気がするのですが、

なんですが、OmuniAuthが渡してくれる`request.env['omniauth.auth']`（`auth_params`はこれを返すだけのメソッド）が
そもそも`Hashie::Mash`なのでテストデータの方でそれに揃えるのは妥当か、（だから↑のように楽しても許されるよね）と思って使いました。

ただまあ、こんな長い説明を要するのを見るに、すっきりしたストラテジーを自作するべきかも知れません……。

kentaro commented on config/locales/translation_ja.yml:
```diff
@@ -124,6 +124,7 @@ ja:
       api: API
       my_page: :helpers.links.my_page
       sign_in: サインイン
+      dev_sign_in: デヴサインイン
```
「開発者としてサインイン」とかにしましょうか><

kentaro commented on config/locales/translation_en.yml:
```diff
@@ -138,6 +138,7 @@ en:
       api: API
       my_page: :helpers.links.my_page
       sign_in: Sign in
+      dev_sign_in: Dev sign in
```
"Sign in as a developer"

kentaro commented:
> POST request to /auth/developer/callback is accepted. Is it OK?

Seems OK. Go ahead.

> No test. Is it needed even though developer strategy is activated only when RAILS_ENV=development?

Do you have any idea how you can test that?

hsbt commented:
> > No test. Is it needed even though developer strategy is activated only when RAILS_ENV=development?
> 
> Do you have any idea how you can test that?

To use `Rails.stub(:env) { "development" }` when running test.

KitaitiMakoto commented:
Naruhodo! Thank you.
The rest of my worry is about database. I'll research and try to resolve it if problems there.

KitaitiMakoto committed:
Dev sign in -> Sign in as a developer

kentaro commented:
> `Rails.stub(:env) { "development" }`

I never came up with such an idea!

KitaitiMakoto committed:
Add spec for developer strategy of OmniAuth

KitaitiMakoto committed:
Sign in as a developer -> Sign in with developer strategy

"Sign in as a developer" sounds like
"Sign in as a developer who has additional privilegeds."

KitaitiMakoto committed:
Fix spec following change of implementation

KitaitiMakoto committed:
Test all about Signin/Signout for development environment

KitaitiMakoto commented:
@kentaro san
I think this branch is ready to be merged.
Could you review it?

hsbt commented on app/contexts/session_context.rb:
```diff
@@ -57,8 +57,8 @@ def update_access_token(credentials)
-  def update_privilege
-    if Settings.github.try(:organizations).present?
+  def update_privilege(provider)
+    if provider == 'github' && Settings.github.try(:organizations).present?
```
To use `Settings.github.organizations.presence` instead of `Settings.github.try(:organizations).present?``

KitaitiMakoto replied:
確かに`try`は全然いりませんね。
`if`節の中では述語の`present?`の方がしっくりくるのですがどうでしょう。

hsbt commented on app/models/user.rb:
```diff
@@ -2,7 +2,7 @@ class User < ActiveRecord::Base
   include LogicallyDeletableRole
 
-  validates :provider, presence: true, inclusion: { in: %w(github) }
+  validates :provider, presence: true, inclusion: { in: Rails.env.development? ? %w(developer github) : %w(github) }
```
Settings に追いやった方が良いか微妙なところですねえ。provider が増えそうなら追いやった方がよさそう。

KitaitiMakoto replied:
おいやります！

hsbt commented on app/views/root/caveat.html.erb:
```diff
@@ -3,7 +3,10 @@
     (Not a member)
   <% end -%>
-  <%= link_to "Sign out", signout_path, method: :delete %>
+  <%= link_to t('layout.nav.sign_out'), signout_path, method: :delete %>
```
pretty good!

hsbt commented on config/routes.rb:
```diff
@@ -3,8 +3,10 @@
+  get    '/dev_signin' => redirect('/auth/developer')
   delete '/signout', to: 'sessions#destroy'
   get    '/auth/:provider/callback', to: 'sessions#create'
+  post   '/auth/developer/callback', to: 'sessions#create' if Rails.env.development?
```
これは不要な気もしますが、無いと動きませんか?

KitaitiMakoto replied:
OmniAuthに最初からついてくるdeveloper strategyだとformのmethodをいじれないぽいんですよね……。

hsbt replied:
nrhd! なら、dev_signout と合わせて development 専用だよという意味で if で囲むとよさそうです。

KitaitiMakoto committed:
Gather routes for dev environment at the same place

kentaro commented:
@KitaitiMakoto
I'll merge this pull request after you do with the points @hsbt said

kentaro commented:
Merge origin/master to this branch, plz.

KitaitiMakoto committed:
Manage OAuth providers in config file

KitaitiMakoto committed:
Remove unneccessary #try

KitaitiMakoto committed:
Resolve conflict

hsbt commented:
コメントが消えていた! `try(:attr).present?` でも許容出来る範囲なのですが、try で存在するかどうかをチェックしてなおかつ present? でチェックというのは何かむずむずするなあと思った次第です。

KitaitiMakoto commented:
> コメントが消えていた!

ソースから該当行が無くなったので「hsbt discussed an outdated diff 2 days ago Show outdated diff」というところに隠れてしまいました。
http://gyazo.com/6560535af460f3231856d27ff9147ce0
（何日か前の @glidenote さんと @kyanny さんのやり取りで知りました）

> 何かむずむずするなあ

指摘されて僕もそう思いました。
24c0dcf6d4802b2543bf59715f08a6a2aa9690c9
↑で`try`だけ外しました。

kentaro commented:
Thanks!!1 Great job!!1
